### PR TITLE
networkd: honour LinkLocalAddressing

### PR DESCRIPTION
**Pull request by ssahani**
fixes systemd#9890

### CODE REVIEW

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -1783,6 +1783,8 @@ int link_up(Link *link) {
 
         log_link_debug(link, "Bringing link up");
 
+        (void) link_ipv6_enabled(link);
```
Comment (yuwata) [inline src/network/networkd-link.c]: Maybe, is this a typo of `link_enable_ipv6()`?

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -1808,41 +1810,39 @@ int link_up(Link *link) {
         if (r < 0)
                 return log_link_error_errno(link, r, "Could not open IFLA_AF_SPEC container: %m");
 
-        if (link_ipv6_enabled(link)) {
-                /* if the kernel lacks ipv6 support setting IFF_UP fails if any ipv6 options are passed */
```
Comment (yuwata) [inline src/network/networkd-link.c]: I think dropping the check is dangerous.

Comment (ssahani) [reply src/network/networkd-link.c]: Actually I dropped I missed that comment.  thanks

File: src/network/networkd-link.c (hunk 1 above)
Comment (ssahani) [reply src/network/networkd-link.c]: we could do that .

**Comment by ssahani** [comment]
I forced pushed a new version

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -284,6 +284,29 @@ static int link_enable_ipv6(Link *link) {
...
+
+        if (link->flags & IFF_LOOPBACK)
+                return 0;
+
+        if (link_ipv6ll_enabled(link))
+                return 0;
+
+        p = strjoina("/proc/sys/net/ipv6/conf/", link->ifname, "/addr_gen_mode");
+
+        r = write_string_file(p, "1", WRITE_STRING_FILE_DISABLE_BUFFER);
```
Comment (yuwata) [inline src/network/networkd-link.c]: In networkd-link.c, we usually also set `WRITE_STRING_FILE_VERIFY_ON_FAILURE`.

Comment (ssahani) [reply src/network/networkd-link.c]: updated thanks !

**Comment by yuwata** [comment]
I mean both WRITE_STRING_FILE_DISABLE_BUFFER and WRITE_STRING_FILE_VERIFY_ON_FAILURE.

**Comment by ssahani** [comment]
ahh right updated.

**Comment by poettering** [comment]
I am not sure I grok this patch? A few lines down from where this code is added it appears to me the same option is set — but via netlink rather than sysctl. i.e. this code will now first turn off ipv6ll via sysctl, and then turn it on again via netlink potentially... This sounds strange to me? What I am not grokking here, can you elaborate?

**Comment by ssahani** [comment]
Yes right but  We not going into the code  of netlink if IPv6 is not enabled. https://github.com/systemd/systemd/blob/80aff27aeb7bbe694395df56be300ff5a4ea4e02/src/network/networkd-link.c#L1809

**Comment by poettering** [comment]
well, but it appears to me that we should have an "else" branch on that if then, which adds the necessary netlink field to turn off the setting without bothering with sysctls. I mean, netlink is always the better choice if both options are available. And if we do the netlink stuff already anywayin one case we definitely should do it for the other case too!

**Comment by ssahani** [comment]
I am not able to figure that out any hint ?

**Comment by ssahani** [comment]
@poettering after hacking the kernel I figure address gen mode and link up does not go together well. So I made it different functions now . 
 
If we put the address genmode with link_up then first time it does not work. but if we bring the device down and restart networkd it does in the second time that's because first time it set the cnf and this time it read from the conf.

**Comment by yuwata** [comment]
Looks better than the previous. I will test this later.

**Comment by yuwata** [comment]
@ssahani LGTM. As usual, I slightly updated your commit (just added one comment), and added tests for `LinkLocalAddressing=`.

@poettering Could you review this? I think this should be included in v241.

**Comment by ssahani** [comment]
Thanks @yuwata .

It seems test is not going well with CentOS CI — . It seems addr_gen_mode does not exist there. So we can test f that file exists then we proceed with the test else we can skip as ususal.
```
ERROR: test_link_local_addressing (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 909, in test_link_local_addressing
    f.readline()
OSError: [Errno 5] Input/output error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 912, in test_link_local_addressing
    self.assertEqual(self.read_ipv6_sysctl_attr('test1', 'addr_gen_mode'), '0')
  File "./test/test-network/systemd-networkd-tests.py", line 114, in read_ipv6_sysctl_attr
    with open(os.path.join(os.path.join(network_sysctl_ipv6_path, link), attribute)) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/proc/sys/net/ipv6/conf/test1/addr_gen_mode'

----------------------------------------------------------------------
Ran 70 tests in 761.081s

FAILED (errors=1, expected failures=7)
[RESULT] FAIL (log file: /root/testsuite-logs.1OU/systemd-networkd-tests.py_FAIL.log)
```

**Comment by yuwata** [comment]
@ssahani Updated. Let's wait for the CIs.

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -1763,6 +1763,84 @@ bool link_has_carrier(Link *link) {
...
+        assert(link);
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 1;
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0)
+                log_link_warning_errno(link, r, "Could not set address genmode for interface: %m");
+
+        return 1;
```
Comment (keszybz) [inline src/network/networkd-link.c:1778]: What is the significance of `1` here? I guess none, so it should be 0 instead.

**Comment by keszybz** [comment]
It fails in TEST-10-ISSUE-2467, with the following
```
[FAILED] Failed to start Testsuite service.
See 'systemctl status testsuite.service' for details.
```
which seems unrelated.

File: src/network/networkd-link.c (hunk 4 above)
Comment (ssahani) [reply src/network/networkd-link.c:1778]: yes I think while we process callback after we get a reply from kernel we return  0 or 1 just to mention the success.

**Pull request merged**